src/IO/Exporter/SpacetimeInterpolator.cpp

```diff
@@ -0,0 +1,206 @@
+                            }},
+                 volume_files_or_glob);
+  const h5::H5File<h5::AccessType::ReadOnly> first_h5file(filenames.front());
+  const auto& first_volfile = first_h5file.get<h5::VolumeData>(subfile_name);
+  const auto all_observation_ids = first_volfile.list_observation_ids();
+  const size_t obs_id =
+      std::visit(SelectObservation{first_volfile}, observation);
+  first_h5file.close();
+  return {std::distance(all_observation_ids.begin(),
+                        std::find(all_observation_ids.begin(),
+                                  all_observation_ids.end(), obs_id)),
+          all_observation_ids.size()};
+}
+
+}  // namespace
+
+template <size_t Dim, typename Frame>
+SpacetimeInterpolator<Dim, Frame>::SpacetimeInterpolator(
+    std::variant<std::vector<std::string>, std::string> volume_files_or_glob,
+    std::string subfile_name, const ObservationVariant& initial_observation,
+    std::vector<std::string> tensor_components, const bool forward_in_time,
+    const size_t num_time_slices)
+    : volume_files_or_glob_(std::move(volume_files_or_glob)),
+      subfile_name_(std::move(subfile_name)),
+      tensor_components_(std::move(tensor_components)),
+      forward_in_time_(forward_in_time),
+      num_time_slices_(num_time_slices) {
+  ASSERT(num_time_slices_ >= time_interpolation_order_ + 1,
+         "The number of time slices must be at least the time interpolation "
+         "order plus one (NumTimeSlices >= "
+             << time_interpolation_order_ + 1 << ").");
+  std::tie(current_obs_step_, num_obs_steps_) = get_observation_step(
+      volume_files_or_glob_, subfile_name_, initial_observation);
+  advance_time();
+}
+
+template <size_t Dim, typename Frame>
+void SpacetimeInterpolator<Dim, Frame>::advance_time() {
+  size_t num_new_slices = num_time_slices_;
+  if (interpolators_.empty()) {
+    // Load `num_new_slices` slices of data, including ghost slices
+    if (forward_in_time_) {
+      current_obs_step_ -= static_cast<int>(num_ghost_slices_);
+    } else {
+      current_obs_step_ += static_cast<int>(num_ghost_slices_);
+    }
+  } else {
+    // Keep last ghost slices in memory and load remaining slices
+    num_new_slices -= time_interpolation_order_;
+    interpolators_.erase(
+        interpolators_.begin(),
+        interpolators_.begin() + static_cast<std::ptrdiff_t>(num_new_slices));
+  }
+  const auto obs_steps_bounds =
+      forward_in_time_ ? std::pair{static_cast<size_t>(current_obs_step_),
+                                   static_cast<size_t>(current_obs_step_) +
+                                       num_new_slices - 1}
+                       : std::pair{static_cast<size_t>(current_obs_step_) -
+                                       num_new_slices + 1,
+                                   static_cast<size_t>(current_obs_step_)};
+  if (obs_steps_bounds.first < 0 or obs_steps_bounds.second >= num_obs_steps_) {
+    ERROR("The requested time steps "
```

Comment:
This should handle reaching the end of the data more gracefully.  Instead of erroring, it should decrease the end time to the maximum allowable value.